File: ha-client/src/executor.ts
/**
 * Prediction executor.
 *
 * Reads the latest prediction JSON from data/predictions/ and executes
 * HVAC control commands via HAClient.
 *
 * Features:
 * - Lazy execution: skips commands when HA is already in the desired state
 * - Override detection: skips devices that have been manually changed since
 *   the last execution (respects user overrides until cleared)
 * - Force mode (--force): bypasses override detection
 */

import { readdir, readFile, writeFile, mkdir } from "node:fs/promises";
import { resolve, dirname } from "node:path";

import type { HAClient, HAEntityState, Prediction } from "./types.ts";
import type { Config } from "./config.ts";
import { config as yamlConfig } from "./yaml-config.ts";

// ── Types ────────────────────────────────────────────────────────────────

interface DeviceState {
  mode?: string;
  target?: number;
}

interface ExecutorState {
  timestamp: string;
  devices: Record<string, DeviceState>;
}

// ── Constants ────────────────────────────────────────────────────────────

/** Ignore temperature differences smaller than this (°F). */
const TARGET_TOLERANCE = 0.5;

/** Convert snake_case to camelCase for prediction JSON key lookup. */
function snakeToCamel(s: string): string {
  const parts = s.split("_");
  return parts[0] + parts.slice(1).map((p) => (p ? p[0]!.toUpperCase() + p.slice(1) : "")).join("");
}

// ── State persistence ────────────────────────────────────────────────────

function executorStatePath(config: Config): string {
  return resolve(config.dataDir, "executor_state.json");
}

async function loadExecutorState(config: Config): Promise<ExecutorState | null> {
  try {
    const content = await readFile(executorStatePath(config), "utf-8");
    return JSON.parse(content) as ExecutorState;
  } catch {
    return null;
  }
}

async function saveExecutorState(config: Config, state: ExecutorState): Promise<void> {
  const p = executorStatePath(config);
  await mkdir(dirname(p), { recursive: true });
  await writeFile(p, JSON.stringify(state, null, 2));
}

// ── HA state readers ─────────────────────────────────────────────────────

function readThermostat(stateMap: Map<string, HAEntityState>, entityId: string): DeviceState {
  const entity = stateMap.get(entityId);
  const target = entity?.attributes["temperature"];
  return { target: typeof target === "number" ? target : undefined };
}

function readMiniSplit(stateMap: Map<string, HAEntityState>, entityId: string): DeviceState {
  const entity = stateMap.get(entityId);
  const target = entity?.attributes["temperature"];
  return {
    mode: entity?.state ?? "unknown",
    target: typeof target === "number" ? target : undefined,
  };
}

function readBlower(stateMap: Map<string, HAEntityState>, entityId: string): DeviceState {
  const entity = stateMap.get(entityId);
  if (!entity || entity.state === "off") return { mode: "off" };
  const preset = entity.attributes["preset_mode"];
  return { mode: typeof preset === "string" ? preset : "low" };
}

// ── Device command helpers ───────────────────────────────────────────────

async function applyThermostat(
  client: HAClient,
  entityId: string,
  target: number,
): Promise<void> {
  await client.callService({
    domain: "climate",
    service: "set_temperature",
    target: { entity_id: entityId },
    serviceData: { temperature: target },
  });
}

async function applyMiniSplit(
  client: HAClient,
  entityId: string,
  mode: string,
  target: number,
): Promise<void> {
  await client.callService({
    domain: "climate",
    service: "set_hvac_mode",
    target: { entity_id: entityId },
    serviceData: { hvac_mode: mode },
  });
  if (mode !== "off") {
    await client.callService({
      domain: "climate",
      service: "set_temperature",
      target: { entity_id: entityId },
      serviceData: { temperature: target },
    });
  }
}

async function applyBlower(
  client: HAClient,
  entityId: string,
  mode: string,
): Promise<void> {
  if (mode === "off") {
    await client.callService({
      domain: "fan",
      service: "turn_off",
      target: { entity_id: entityId },
    });
  } else {
    await client.callService({
      domain: "fan",
      service: "turn_on",
      target: { entity_id: entityId },
    });
    await client.callService({
      domain: "fan",
      service: "set_preset_mode",
      target: { entity_id: entityId },
      serviceData: { preset_mode: mode },
    });
  }
}

// ── Prediction loader ────────────────────────────────────────────────────

async function loadLatestPrediction(predictionsDir: string): Promise<Prediction | null> {
  let files: string[];
  try {
    files = await readdir(predictionsDir);
  } catch {
    return null;
  }
  const jsonFiles = files.filter((f) => f.startsWith("command_") && f.endsWith(".json")).sort();
  const latest = jsonFiles[jsonFiles.length - 1];
  if (!latest) return null;
  const content = await readFile(resolve(predictionsDir, latest), "utf-8");
  return JSON.parse(content) as Prediction;
}

// ── Main executor ────────────────────────────────────────────────────────

export async function executePrediction(
  client: HAClient,
  config: Config,
  force: boolean = false,
): Promise<void> {
  const prediction = await loadLatestPrediction(config.predictionsDir);
  if (!prediction) {
    console.log("[executor] No prediction found, skipping");
    return;
  }

  console.log(
    `[executor] Executing prediction from ${prediction.timestamp}`
    + ` (confidence: ${prediction.confidence.toFixed(2)})`,
  );
  if (force) console.log("[executor] Force mode — overrides will be ignored");

  // Fetch current HA states for all controlled entities
  const controlledEntities = Object.values(yamlConfig.effectors).map((e) => e.entityId);
  const states = await client.getStates(controlledEntities);
  const stateMap = new Map<string, HAEntityState>(states.map((s) => [s.entity_id, s]));

  // Load last-applied state (what the executor actually set last time)
  const lastState = await loadExecutorState(config);
  const OVERRIDE_STALE_MS = 30 * 60 * 1000; // 30 minutes (2× loop interval)

  let last = lastState?.devices ?? {};
  if (lastState?.timestamp) {
    const age = Date.now() - new Date(lastState.timestamp).getTime();
    if (age > OVERRIDE_STALE_MS) {
      console.log(
        `[executor] Executor state is ${Math.round(age / 60000)}m old — clearing overrides`,
      );
      last = {};
    }
  }

  // Clone last-applied — updated per-device only when we actually apply
  const updated: Record<string, DeviceState> = { ...last };

  let applied = 0;
  let lazySkips = 0;
  let overrideSkips = 0;

  // ── Iterate all effectors from config ─────────────────────────────────

  for (const [name, ecfg] of Object.entries(yamlConfig.effectors)) {
    if (ecfg.domain === "climate" && ecfg.modeControl === "manual") {
      // ── Manual-mode climate (thermostat): target only ───────────────
      const targetKey = snakeToCamel(`${name}_target`);
      const desiredTarget = prediction[targetKey] as number | undefined;

      if (desiredTarget === undefined) {
        console.log(`[executor] ${name}: ineligible, skipping`);
        continue;
      }

      const current = readThermostat(stateMap, ecfg.entityId);

      if (current.target !== undefined && Math.abs(current.target - desiredTarget) < TARGET_TOLERANCE) {
        console.log(`[executor] ${name}: already at ${current.target}°F, skipping`);
        lazySkips++;
        continue;
      }

      const lastApplied = last[name];
      if (
        !force
        && lastApplied?.target !== undefined
        && current.target !== undefined
        && Math.abs(current.target - lastApplied.target) > TARGET_TOLERANCE
      ) {
        console.log(
          `[executor] ${name}: override detected`
          + ` (current ${current.target}°F, we last set ${lastApplied.target}°F), skipping`,
        );
        overrideSkips++;
        continue;
      }

      await applyThermostat(client, ecfg.entityId, desiredTarget);
      console.log(`[executor] ${name}: set to ${desiredTarget}°F`);
      updated[name] = { target: desiredTarget };
      applied++;

    } else if (ecfg.domain === "climate") {
      // ── Automatic-mode climate (mini-split): mode + target ──────────
      const modeKey = snakeToCamel(`${name}_mode`);
      const targetKey = snakeToCamel(`${name}_target`);
      const desiredMode = prediction[modeKey] as string | undefined;
      const desiredTarget = prediction[targetKey] as number | undefined;

      if (desiredMode === undefined) {
        console.log(`[executor] ${name}: not in command, skipping`);
        continue;
      }

      const current = readMiniSplit(stateMap, ecfg.entityId);

      const modeMatch = current.mode === desiredMode;
      const targetMatch = desiredMode === "off"
        || (desiredTarget !== undefined && current.target !== undefined
          && Math.abs(current.target - desiredTarget) < TARGET_TOLERANCE);
      if (modeMatch && targetMatch) {
        const desc = desiredMode === "off" ? "off" : `${desiredMode} @ ${current.target}°F`;
        console.log(`[executor] ${name}: already ${desc}, skipping`);
        lazySkips++;
        continue;
      }

      const lastApplied = last[name];
      if (!force && lastApplied?.mode !== undefined && current.mode !== lastApplied.mode) {
        console.log(
          `[executor] ${name}: override detected`
          + ` (currently ${current.mode}, we last set ${lastApplied.mode}), skipping`,
        );
        overrideSkips++;
        continue;
      }

      await applyMiniSplit(client, ecfg.entityId, desiredMode, desiredTarget ?? 72);
      const desc = desiredMode === "off" ? "off" : `${desiredMode} @ ${desiredTarget}°F`;
      console.log(`[executor] ${name}: set to ${desc}`);
      updated[name] = { mode: desiredMode, target: desiredTarget };
      applied++;

    } else {
      // ── Fan entity (blower): mode only ─────────────────────────────
      const modeKey = snakeToCamel(`${name}_mode`);
      const desiredMode = prediction[modeKey] as string | undefined;

      if (desiredMode === undefined) {
        console.log(`[executor] ${name}: not in command, skipping`);
        continue;
      }

      const current = readBlower(stateMap, ecfg.entityId);

      if (current.mode === desiredMode) {
        console.log(`[executor] ${name}: already ${desiredMode}, skipping`);
        lazySkips++;
        continue;
      }

      const lastApplied = last[name];
      if (!force && lastApplied?.mode !== undefined && current.mode !== lastApplied.mode) {
        console.log(
          `[executor] ${name}: override detected`
          + ` (currently ${current.mode}, we last set ${lastApplied.mode}), skipping`,
        );
        overrideSkips++;
        continue;
      }

      await applyBlower(client, ecfg.entityId, desiredMode);
      console.log(`[executor] ${name}: set to ${desiredMode}`);
      updated[name] = { mode: desiredMode };
      applied++;
    }
  }

  // ── Save state ─────────────────────────────────────────────────────────

  await saveExecutorState(config, {
    timestamp: new Date().toISOString(),
    devices: updated,
  });

  console.log(
    `[executor] Done: ${applied} applied, ${lazySkips} already correct,`
    + ` ${overrideSkips} overrides respected`,
  );
}

File: src/weatherstat/executor.py
"""HVAC command executor — applies control decisions to Home Assistant.

Reads the latest command JSON from predictions/, compares desired state against
current HA entity state, and issues service calls for any changes.

Features:
- Lazy execution: skips commands when HA is already in the desired state
- Override detection: skips devices whose state has been manually changed
  since the last execution (respects user overrides until cleared)
- Force mode: bypasses override detection
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from datetime import UTC, datetime

import requests

from weatherstat.config import DATA_DIR, EFFECTORS, PREDICTIONS_DIR

TARGET_TOLERANCE = 0.5
OVERRIDE_STALE_SECONDS = 30 * 60  # 30 minutes

_EXECUTOR_STATE_FILE = DATA_DIR / "executor_state.json"


# ── Types ───────────────────────────────────────────────────────────────────


@dataclass
class DeviceAction:
    """Result of attempting to execute a command for one device."""

    name: str
    action: str  # "applied", "already_correct", "override", "ineligible"
    detail: str = ""


@dataclass
class ExecutorResult:
    """Summary of an execution run."""

    actions: list[DeviceAction] = field(default_factory=list)

    @property
    def applied(self) -> int:
        return sum(1 for a in self.actions if a.action == "applied")

    @property
    def already_correct(self) -> int:
        return sum(1 for a in self.actions if a.action == "already_correct")

    @property
    def overrides(self) -> dict[str, str]:
        """effector_name -> detail for devices with detected overrides."""
        return {a.name: a.detail for a in self.actions if a.action == "override"}


# ── HA helpers ──────────────────────────────────────────────────────────────


def _ha_headers() -> dict[str, str]:
    return {
        "Authorization": f"Bearer {os.environ.get('HA_TOKEN', '')}",
        "Content-Type": "application/json",
    }


def _ha_url() -> str:
    return os.environ.get("HA_URL", "")


def _get_entity_state(entity_id: str) -> dict | None:
    """Fetch a single entity's state from HA REST API."""
    url = _ha_url()
    if not url:
        return None
    try:
        resp = requests.get(f"{url}/api/states/{entity_id}", headers=_ha_headers(), timeout=10)
        if resp.status_code == 200:
            return resp.json()
    except requests.RequestException:
        pass
    return None


def _call_service(domain: str, service: str, entity_id: str, data: dict | None = None) -> bool:
    """Call an HA service. Returns True on success."""
    url = _ha_url()
    if not url:
        return False
    payload: dict = {"entity_id": entity_id}
    if data:
        payload.update(data)
    try:
        resp = requests.post(
            f"{url}/api/services/{domain}/{service}",
            headers=_ha_headers(),
            json=payload,
            timeout=10,
        )
        return resp.status_code == 200
    except requests.RequestException:
        return False


# ── HA state readers ────────────────────────────────────────────────────────


def _read_climate_target(entity_id: str) -> float | None:
    """Read climate entity's target temperature."""
    state = _get_entity_state(entity_id)
    if not state:
        return None
    target = state.get("attributes", {}).get("temperature")
    return float(target) if isinstance(target, (int, float)) else None


def _read_climate_mode(entity_id: str) -> tuple[str | None, float | None]:
    """Read climate entity's mode and target. Returns (mode, target)."""
    state = _get_entity_state(entity_id)
    if not state:
        return None, None
    mode = state.get("state")
    target = state.get("attributes", {}).get("temperature")
    return mode, float(target) if isinstance(target, (int, float)) else None


def _read_fan_mode(entity_id: str) -> str | None:
    """Read fan entity's effective mode (off / preset_mode)."""
    state = _get_entity_state(entity_id)
    if not state:
        return None
    if state.get("state") == "off":
        return "off"
    preset = state.get("attributes", {}).get("preset_mode")
    return str(preset) if preset else "low"


# ── Executor state persistence ──────────────────────────────────────────────


def _load_executor_state() -> dict:
    if _EXECUTOR_STATE_FILE.exists():
        try:
            return json.loads(_EXECUTOR_STATE_FILE.read_text())
        except (json.JSONDecodeError, OSError):
            pass
    return {}


def _save_executor_state(devices: dict[str, dict]) -> None:
    state = {"timestamp": datetime.now(UTC).isoformat(), "devices": devices}
    _EXECUTOR_STATE_FILE.write_text(json.dumps(state, indent=2))


# ── Command loader ──────────────────────────────────────────────────────────


def _load_latest_command() -> dict | None:
    """Load the most recent command_*.json from predictions/."""
    if not PREDICTIONS_DIR.exists():
        return None
    files = sorted(PREDICTIONS_DIR.glob("command_*.json"))
    if not files:
        return None
    try:
        return json.loads(files[-1].read_text())
    except (json.JSONDecodeError, OSError):
        return None


def _snake_to_camel(s: str) -> str:
    parts = s.split("_")
    return parts[0] + "".join(p.capitalize() for p in parts[1:])


# ── Main executor ───────────────────────────────────────────────────────────


def execute(*, force: bool = False, log: print = print) -> ExecutorResult:
    """Execute the latest command JSON against Home Assistant.

    Args:
        force: If True, ignore detected overrides.
        log: Callable for status messages (default: print).

    Returns:
        ExecutorResult with per-device actions.
    """
    result = ExecutorResult()

    command = _load_latest_command()
    if not command:
        log("[executor] No command found")
        return result

    ts = command.get("timestamp", "?")
    confidence = command.get("confidence", 0)
    log(f"[executor] Executing command from {ts} (confidence: {confidence:.2f})")
    if force:
        log("[executor] Force mode — overrides will be ignored")

    # Load last-applied state
    exec_state = _load_executor_state()
    last_devices: dict[str, dict] = exec_state.get("devices", {})

    # Check staleness
    state_ts = exec_state.get("timestamp", "")
    if state_ts:
        try:
            dt = datetime.fromisoformat(state_ts.replace("Z", "+00:00"))
            age = (datetime.now(UTC) - dt).total_seconds()
            if age > OVERRIDE_STALE_SECONDS:
                log(f"[executor] Executor state is {int(age / 60)}m old — clearing overrides")
                last_devices = {}
        except (ValueError, TypeError):
            pass

    updated_devices = dict(last_devices)

    for eff in EFFECTORS:
        if eff.control_type in ("trajectory",) and eff.mode_control == "manual":
            # Manual-mode climate (thermostat): target only
            target_key = _snake_to_camel(f"{eff.name}_target")
            desired_target = command.get(target_key)
            if desired_target is None:
                result.actions.append(DeviceAction(eff.name, "ineligible"))
                continue

            desired_target = float(desired_target)
            current_target = _read_climate_target(eff.entity_id)

            if current_target is not None and abs(current_target - desired_target) < TARGET_TOLERANCE:
                log(f"[executor] {eff.name}: already at {current_target}°F")
                result.actions.append(DeviceAction(eff.name, "already_correct", f"{current_target}°F"))
                continue

            last = last_devices.get(eff.name, {})
            last_target = last.get("target")
            if (
                not force
                and last_target is not None
                and current_target is not None
                and abs(current_target - last_target) > TARGET_TOLERANCE
            ):
                detail = f"target {current_target}°F, we set {last_target}°F"
                log(f"[executor] {eff.name}: override detected ({detail})")
                result.actions.append(DeviceAction(eff.name, "override", detail))
                continue

            _call_service("climate", "set_temperature", eff.entity_id, {"temperature": desired_target})
            log(f"[executor] {eff.name}: set to {desired_target}°F")
            updated_devices[eff.name] = {"target": desired_target}
            result.actions.append(DeviceAction(eff.name, "applied", f"{desired_target}°F"))

        elif eff.control_type == "regulating":
            # Automatic-mode climate (mini-split): mode + target
            mode_key = _snake_to_camel(f"{eff.name}_mode")
            target_key = _snake_to_camel(f"{eff.name}_target")
            desired_mode = command.get(mode_key)
            desired_target = command.get(target_key)

            if desired_mode is None:
                result.actions.append(DeviceAction(eff.name, "ineligible"))
                continue

            current_mode, current_target = _read_climate_mode(eff.entity_id)

            mode_match = current_mode == desired_mode
            target_match = (
                desired_mode == "off"
                or (desired_target is not None and current_target is not None and abs(current_target - float(desired_target)) < TARGET_TOLERANCE)
            )
            if mode_match and target_match:
                desc = "off" if desired_mode == "off" else f"{desired_mode} @ {current_target}°F"
                log(f"[executor] {eff.name}: already {desc}")
                result.actions.append(DeviceAction(eff.name, "already_correct", desc))
                continue

            last = last_devices.get(eff.name, {})
            if not force and last.get("mode") is not None and current_mode != last["mode"]:
                detail = f"mode {current_mode}, we set {last['mode']}"
                log(f"[executor] {eff.name}: override detected ({detail})")
                result.actions.append(DeviceAction(eff.name, "override", detail))
                continue

            _call_service("climate", "set_hvac_mode", eff.entity_id, {"hvac_mode": desired_mode})
            if desired_mode != "off" and desired_target is not None:
                _call_service("climate", "set_temperature", eff.entity_id, {"temperature": float(desired_target)})
            desc = "off" if desired_mode == "off" else f"{desired_mode} @ {desired_target}°F"
            log(f"[executor] {eff.name}: set to {desc}")
            updated_devices[eff.name] = {"mode": desired_mode, "target": desired_target}
            result.actions.append(DeviceAction(eff.name, "applied", desc))

        elif eff.control_type == "binary":
            # Fan entity (blower): mode only
            mode_key = _snake_to_camel(f"{eff.name}_mode")
            desired_mode = command.get(mode_key)

            if desired_mode is None:
                result.actions.append(DeviceAction(eff.name, "ineligible"))
                continue

            current_mode = _read_fan_mode(eff.entity_id)

            if current_mode == desired_mode:
                log(f"[executor] {eff.name}: already {desired_mode}")
                result.actions.append(DeviceAction(eff.name, "already_correct", desired_mode))
                continue

            last = last_devices.get(eff.name, {})
            if not force and last.get("mode") is not None and current_mode != last["mode"]:
                detail = f"mode {current_mode}, we set {last['mode']}"
                log(f"[executor] {eff.name}: override detected ({detail})")
                result.actions.append(DeviceAction(eff.name, "override", detail))
                continue

            if desired_mode == "off":
                _call_service("fan", "turn_off", eff.entity_id)
            else:
                _call_service("fan", "turn_on", eff.entity_id)
                _call_service("fan", "set_preset_mode", eff.entity_id, {"preset_mode": desired_mode})
            log(f"[executor] {eff.name}: set to {desired_mode}")
            updated_devices[eff.name] = {"mode": desired_mode}
            result.actions.append(DeviceAction(eff.name, "applied", desired_mode))

    _save_executor_state(updated_devices)

    log(f"[executor] Done: {result.applied} applied, {result.already_correct} correct, {len(result.overrides)} overrides")
    return result


def main() -> None:
    import argparse

    parser = argparse.ArgumentParser(description="Execute latest HVAC command via Home Assistant")
    parser.add_argument("--force", action="store_true", help="Ignore manual overrides")
    args = parser.parse_args()
    execute(force=args.force)


if __name__ == "__main__":
    main()
